cleverhans/attacks_tf.py

```diff
@@ -726,3 +726,126 @@ def compare(x, y):
+    :param max_iter: Maximum number of iteration for deepfool
+    :param clip_min: Minimum value for components of the example returned
+    :param clip_max: Maximum value for components of the example returned
+    :param nb_classes: Number of model output classes
+    :return: Adversarial examples
+    """
+
+    X_adv = np.zeros(X.shape)
+    X_adv = deepfool_attack(sess, x, pred, logits, grads, X, nb_candidate,
+                            overshoot, max_iter, clip_min, clip_max, feed=feed)
+
+    return np.asarray(X_adv, dtype=np.float32)
+
+
+def deepfool_attack(sess, x, predictions, logits, grads, sample, nb_candidate,
+                    overshoot, max_iter, clip_min, clip_max, feed=None):
+    """
+    TensorFlow implementation of the deepfool.
+    Paper link: see https://arxiv.org/pdf/1511.04599.pdf
+    :param sess: TF session
+    :param x: The input placeholder
+    :param predictions: The model's symbolic output (linear output,
+        pre-softmax)
+    :param logits: The model's unnormalized output tensor (the input to
+                   the softmax layer)
+    :param grads: Symbolic gradients
+    :param sample: Numpy array with sample input
+    :param nb_candidate: The number of adversarial class candidates
+    :param overshoot: A termination criterion to prevent vanishing updates
+    :param max_iter: Maximum number of iteration for deepfool
+    :param clip_min: Minimum value for components of the example returned
+    :param clip_max: Maximum value for components of the example returned
+    :return: an adversarial sample
+    """
+    import copy
+
+    adv_x = copy.copy(sample)
+    # Initialize the loop variables
+    iteration = 0
+    current = utils_tf.model_argmax(sess, x, logits, adv_x, feed=feed)
+    if current.shape == ():
+        current = np.array([current])
+    w = np.squeeze(np.zeros(sample.shape[1:4]))  # same shape as original image
+    r_tot = np.zeros(sample.shape)
+    original = current  # use original label as the reference
+
+    _logger.debug("Starting DeepFool attack up to {} iterations".
+                  format(max_iter))
+    # Repeat this main loop until we have achieved misclassification
+    while (np.any(current == original) and iteration < max_iter):
+
+        if iteration % 5 == 0 and iteration > 0:
+            _logger.info("Attack result at iteration {} is {}".format(
+                iteration,
+                current))
+        gradients = sess.run(grads, feed_dict={x: adv_x})
+        f = sess.run(predictions, feed_dict={x: adv_x})
+        for idx in range(sample.shape[0]):
+            pert = np.inf
+            if current[idx] != original[idx]:
+                continue
+            for k in range(1, nb_candidate):
+                w_k = gradients[idx, k, ...] - gradients[idx, 0, ...]
+                f_k = f[idx, k] - f[idx, 0]
+                pert_k = (abs(f_k) + 0.00001)/np.linalg.norm(w_k.flatten())
```

Comment:
Can you add spaces around the `/` operator?

Reply:
Is this the exact algorithm used in the paper? If no, where is the value `0.00001` coming from?